x/emissions/keeper/actor_utils/losses.go

```diff
@@ -69,44 +63,22 @@ func CloseReputerNonce(
 		return err
 	}
 
-	reputerLossBundles, err := k.GetReputerLossBundlesAtBlock(ctx, topicId, nonce.BlockHeight)
+	// Get active reputers for the topic
+	activeReputerAddresses, err := k.GetActiveReputersForTopic(ctx, topic.Id)
 	if err != nil {
-		return types.ErrNoValidBundles
+		return err
 	}
 
-	// Iterate through the array to ensure each reputer is in the whitelist
-	// and get get score for each reputer => later we can skim only the top few by score descending
 	lossBundlesByReputer := make([]*types.ReputerValueBundle, 0)
 	stakesByReputer := make(map[string]cosmosMath.Int)
-	for _, bundle := range reputerLossBundles.ReputerValueBundles {
-		err := bundle.Validate()
+	for _, address := range activeReputerAddresses {
+		bundle, err := k.GetReputerLatestLossByTopicId(ctx, topic.Id, address)
 		if err != nil {
-			continue
-		}
-
-		reputer := bundle.ValueBundle.Reputer
-
-		// Check that the reputer's value bundle is for a topic matching the given topic
-		if bundle.ValueBundle.TopicId != topicId {
-			continue
-		}
-		// Check that the reputer's value bundle is for a nonce matching the given nonce
-		if bundle.ValueBundle.ReputerRequestNonce.ReputerNonce.BlockHeight != nonce.BlockHeight {
-			continue
-		}
-
-		// Check that the reputer is registered in the topic
-		isReputerRegistered, err := k.IsReputerRegisteredInTopic(ctx, bundle.ValueBundle.TopicId, reputer)
-		if err != nil {
-			continue
-		}
-		// We'll keep what we can get from the payload, but we'll ignore the rest
-		if !isReputerRegistered {
-			continue
+			return types.ErrNoValidBundles
```

Comment:
imho they are not that redundant.

ingress checks and closeReputerNonce happen at different times, and react (error) on different ways in each case(to user, or to validator logs); and for completeness, none of the function should allow that.

Also, it is possible that the reputer has deregistered during the time from the tx was made to the time when the reputer nonce is closed.

Reply:
Not all functions need to check everything -- a function's "contract" it signs with the world includes its intention specified in its comments (by its declaration). That can include making assumptions about the arguments and viewed state. This is esp true and necessary in a compute-constrained environment.

We're also building `x/emissions` as nearly a monolith, so we get to make more "assumptions" like this (as opposed to a modular/microservice architecture).

The real pull to add these validations is if we expect / should prepare to ever edit the values in these KV stores at any point between ingress and their usage. This is minimized by adding these validations to within the scope of any setter. But then again such checks too can be averted by state-editing migrations or not using the designated setters.

So I'm a bit torn but reintroducing them seems unavoidable if we want to max safety.

Curious @spooktheducks thoughts...

Reply:
`d e f e n s e   i n   d e p t h`

Reply:
There is no deregistration. The checks are made earlier. I'd like to talk more offline ab it -- can quickly add back in if we feel we want to. Will merge the PR now though as it looks good and it does so much more than touch this validation.